samza-api/src/main/java/org/apache/samza/operators/descriptors/GenericSystemDescriptor.java

```diff
@@ -0,0 +1,136 @@
+ * Additional system specific properties may be provided using {@link #withSystemConfigs}
+ * <p>
+ * System properties configured using a descriptor override corresponding properties provided in configuration.
+ *
+ * @param <SystemMessageType> type of messages in this system
+ */
+@SuppressWarnings("unchecked")
+public final class GenericSystemDescriptor<SystemMessageType>
+    extends SimpleSystemDescriptor<SystemMessageType, GenericSystemDescriptor<SystemMessageType>> {
+
+  /**
+   * Constructs a {@link GenericSystemDescriptor} instance with no system level serde.
+   * Serdes must be provided explicitly at stream level when getting input or output descriptors.
+   *
+   * @param systemName name of this system
+   * @param factoryClassName name of the SystemFactory class for this system
+   */
+  public GenericSystemDescriptor(String systemName, String factoryClassName) {
+    super(systemName, factoryClassName, null);
+  }
+
+  /**
+   * Constructs a {@link GenericSystemDescriptor} instance with {@code systemSerde} as the system level serde.
+   *
+   * @param systemName name of this system
+   * @param factoryClassName name of the SystemFactory class for this system
+   * @param systemSerde default serde for the system, or null.
+   *                    If null, input/output descriptor serde must be provided at a stream level.
+   *                    A {@code KVSerde<NoOpSerde, NoOpSerde>} or {@code NoOpSerde} may be provided if the
+   *                    System's consumer deserializes the incoming messages itself, and no further deserialization
+   *                    is required from the framework.
+   */
+  public GenericSystemDescriptor(String systemName, String factoryClassName, Serde<SystemMessageType> systemSerde) {
+    super(systemName, factoryClassName, systemSerde);
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public GenericInputDescriptor<SystemMessageType> getInputDescriptor(String streamId) {
+    if (!getSystemSerde().isPresent()) {
+      throw new SamzaException(
+          String.format("System: %s does not have a system level serde. " +
+              "Serde for input stream: %s must be specified explicitly, e.g., using " +
+              "GenericSystemDescriptor#getInputDescriptor(String, Serde)", getSystemName(), streamId));
+    }
+    return new GenericInputDescriptor<>(streamId, this, null, getSystemSerde().get());
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public <StreamMessageType> GenericInputDescriptor<StreamMessageType> getInputDescriptor(
+      String streamId, Serde<StreamMessageType> serde) {
+    return new GenericInputDescriptor<>(streamId, this, null, serde);
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public <StreamMessageType> GenericInputDescriptor<StreamMessageType> getInputDescriptor(
+      String streamId, InputTransformer<StreamMessageType> transformer) {
```

Comment:
I was planning to remove them after we have a consistent initial API. Now that we only have 2 getInputDescriptor variants, do you still feel strongly about leaving them in?

Reply:
I don't feel as strongly, but I also still prefer removal after we have the consistent initial API, since we don't have a concrete use case. The simpler the API, the better, and it means less code to support until we know we really need it. Also, it could turn out that the real future use case needs a different pattern/signature anyways.